apps/src/gamelab/AnimationPicker/animationPickerModule.js

```diff
@@ -110,27 +110,34 @@ export function handleUploadComplete(result) {
     const { goal, uploadFilename } = getState().animationPicker;
     const key = result.filename.replace(/\.png$/i, '');
     const sourceUrl = animationsApi.basePath(key + '.png');
-
-    loadImageMetadata(sourceUrl, metadata => {
-      const animation = _.assign({}, metadata, {
-        name: uploadFilename,
-        sourceUrl: sourceUrl,
-        size: result.size,
-        version: result.versionId
-      });
-
-      if (goal === Goal.NEW_ANIMATION) {
-        dispatch(addAnimation(key, animation));
-      } else if (goal === Goal.NEW_FRAME) {
-        dispatch(appendCustomFrames(animation));
-      }
-      dispatch(hide());
-    }, () => {
+    const onImageMetadataLoaded = buildOnImageMetadataLoaded(uploadFilename, goal, key, result, dispatch);
```

Comment:
Reviewing now, I'm questioning our choice of arguments to this method:

- `key` is derived from `result` and its only use is to be passed into this method, which also receives `result`.  Maybe we should compute `key` within the `buildOnImageMetadataLoaded` function. 
- Likewise, `uploadFilename` and `goal` aren't actually needed in this method and they should still be in the state when the `onImageMetadataLoaded` callback fires.

It's possible want we really want here is to call
```js
buildOnImageMetadataLoaded(dispatch, getState, result);
```
which echoes the signature and [redux-thunk](https://github.com/gaearon/redux-thunk) return value of `handleUploadComplete`, and then derive the other values in our helper.  In fact, that might make it easier to test that we derive those values correctly. 

It does mean you'd need to make sure the `animationPicker` module's redux state is set up correctly in your unit test.

Reply:
Moving the look-up of goal and uploadFilename makes a lot of sense for organization and I believe the test already supports this state-wise, so that'll be quick and easy to change.

For 'key', since it's used to look-up the sourceUrl in handleUploadComplete, I think it makes more sense to still compute it there and pass it into buildOnImageMetadataLoaded. Thoughts?

Reply:
:+1: My bad, missed the transitive dependency there.